src/pybind/mgr/dashboard/services/access_control.py

```diff
@@ -0,0 +1,617 @@
+    Scope.GLOBAL: [_P.READ, _P.CREATE, _P.UPDATE, _P.DELETE],
+    Scope.RBD_IMAGE: [_P.READ, _P.CREATE, _P.UPDATE, _P.DELETE],
+    Scope.ISCSI: [_P.READ, _P.CREATE, _P.UPDATE, _P.DELETE],
+    Scope.RBD_MIRRORING: [_P.READ, _P.CREATE, _P.UPDATE, _P.DELETE],
+})
+
+
+# RadosGW manager role provides all permissions for block related scopes
+RGW_MGR_ROLE = Role('RGW Manager', {
+    Scope.GLOBAL: [_P.READ, _P.CREATE, _P.UPDATE, _P.DELETE],
+    Scope.RGW: [_P.READ, _P.CREATE, _P.UPDATE, _P.DELETE],
+})
+
+
+# Cluster manager role provides all permission for OSDs, Monitors, and
+# Config options
+CLUSTER_MGR_ROLE = Role('Cluster Manager', {
+    Scope.GLOBAL: [_P.READ, _P.CREATE, _P.UPDATE, _P.DELETE],
+    Scope.HOSTS: [_P.READ, _P.CREATE, _P.UPDATE, _P.DELETE],
+    Scope.OSD: [_P.READ, _P.CREATE, _P.UPDATE, _P.DELETE],
+    Scope.MONITOR: [_P.READ, _P.CREATE, _P.UPDATE, _P.DELETE],
+    Scope.MANAGER: [_P.READ, _P.CREATE, _P.UPDATE, _P.DELETE],
+    Scope.CONFIG_OPT: [_P.READ, _P.CREATE, _P.UPDATE, _P.DELETE],
+})
+
+
+# Pool manager role provides all permissions for pool related scopes
+POOL_MGR_ROLE = Role('Pool Manager', {
+    Scope.GLOBAL: [_P.READ, _P.CREATE, _P.UPDATE, _P.DELETE],
+    Scope.POOL: [_P.READ, _P.CREATE, _P.UPDATE, _P.DELETE],
+})
+
+# Pool manager role provides all permissions for CephFS related scopes
+CEPHFS_MGR_ROLE = Role('CephFS Manager', {
+    Scope.GLOBAL: [_P.READ, _P.CREATE, _P.UPDATE, _P.DELETE],
+    Scope.CEPHFS: [_P.READ, _P.CREATE, _P.UPDATE, _P.DELETE],
+})
+
+
+SYSTEM_ROLES = {
+    ADMIN_ROLE.name: ADMIN_ROLE,
+    READ_ONLY_ROLE.name: READ_ONLY_ROLE,
+    BLOCK_MGR_ROLE.name: BLOCK_MGR_ROLE,
+    RGW_MGR_ROLE.name: RGW_MGR_ROLE,
+    CLUSTER_MGR_ROLE.name: CLUSTER_MGR_ROLE,
+    POOL_MGR_ROLE.name: POOL_MGR_ROLE,
+    CEPHFS_MGR_ROLE.name: CEPHFS_MGR_ROLE,
+}
+
+
+class User(object):
+    def __init__(self, username, password, name=None, email=None, roles=None):
+        self.username = username
+        self.password = password
+        self.name = name
+        self.email = email
+        if roles is None:
+            self.roles = set()
+        else:
+            self.roles = roles
+
+    def set_roles(self, roles):
+        self.roles = set(roles)
+
+    def add_roles(self, roles):
+        self.roles = self.roles.union(set(roles))
+
+    def del_roles(self, roles):
+        for role in roles:
+            if role not in self.roles:
+                raise RoleNotInUser(role.name, self.username)
+        self.roles.difference_update(set(roles))
+
+    def authorize(self, scope, permissions):
+        for role in self.roles:
+            if role.authorize(scope, permissions):
+                return True
+        return False
+
+    def permissions_dict(self):
+        perms = {}
+        for role in self.roles:
+            for scope, perms_list in role.scopes_permissions.items():
+                if scope in perms:
+                    perms_tmp = set(perms[scope]) + set(perms_list)
```

Comment:
I'm getting the following error while trying to login with a user that has more than one role:

`TypeError: unsupported operand type(s) for +: \'set\' and \'set\'`

How to reproduce:

1. `bin/ceph dashboard ac-user-create test test`
2. `bin/ceph dashboard ac-user-add-roles test Administrator`
3. `bin/ceph dashboard ac-user-add-roles test Read-Only`
4. Login with user `test`